### PR TITLE
Get and Set Globals

### PR DESCRIPTION
**Pull request by RoyalIcing**
Add `Wasmex.Instance.get_global_value/3` and `Wasmex.Instance.set_global_value/4` to read and write globals.

I don’t believe languages like Rust or C allow exporting globals, but in raw WebAssembly this is very much possible:

```wasm
(module
  (global (export "meaning_of_life") i32 (i32.const 42))
  (global (export "count") (mut i64) (i64.const 0))
)
```

I’ve followed what appears to be the convention of returning string errors. If you like I can make them atoms like `{:error, :not_found}` or `{:error, :immutable}`.

### CODE REVIEW

File: lib/wasmex/instance.ex [hunk 1]
```diff
@@ -172,6 +172,47 @@ defmodule Wasmex.Instance do
   def memory(store, instance) do
     Wasmex.Memory.from_instance(store, instance)
   end
+
+  @doc ~S"""
+  Reads the value of an exported global.
```
Comment (RoyalIcing) [inline lib/wasmex/instance.ex:177]: Do you want doctests here?

File: native/wasmex/src/instance.rs [hunk 2]
```diff
@@ -225,6 +316,36 @@ pub enum WasmValue {
     F64(f64),
 }
 
+fn decode_term_as_wasm_value(expected_type: ValType, term: Term) -> Option<WasmValue> {
```
Comment (RoyalIcing) [inline native/wasmex/src/instance.rs:319]: I’ve copied and extract this from below. I could easily DRY this up by making the other use this function.

File: lib/wasmex/instance.ex (hunk 1 above)
Comment (RoyalIcing) [reply lib/wasmex/instance.ex:177]: I added some doctests

File: native/wasmex/src/instance.rs (hunk 2 above)
Comment (RoyalIcing) [reply native/wasmex/src/instance.rs:322]: I reused this.

File: lib/wasmex/instance.ex (hunk 1 above)
Comment (tessi) [reply lib/wasmex/instance.ex:177]: These doctests are great! Thanks for adding them

File: test/example_wasm_files/globals.wat [hunk 3]
```diff
@@ -0,0 +1,7 @@
+(module
+  (global $meaning_of_life (export "meaning_of_life") i32 (i32.const 42))
+  (global (export "count_32") (mut i32) (i32.const -32))
+  (global (export "count_64") (mut i64) (i64.const -64))
+  (global (export "bad_pi_32") (mut f32) (f32.const 0))
+  (global (export "bad_pi_64") (mut f64) (f64.const 0))
```
Comment (tessi) [inline test/example_wasm_files/globals.wat:6]: 😅 haha, this really is a bad pi

File: test/wasmex/instance_test.exs [hunk 4]
```diff
@@ -195,4 +195,58 @@ defmodule Wasmex.InstanceTest do
...
+    test t(&Wasmex.Instance.get_global_value/3), context do
+      store = context[:store]
+      instance = context[:instance]
+
+      assert {:error, "exported global `unknown_global` not found"} =
+               Wasmex.Instance.get_global_value(store, instance, "unknown_global")
+
+      assert {:ok, 42} = Wasmex.Instance.get_global_value(store, instance, "meaning_of_life")
+      assert {:ok, -32} = Wasmex.Instance.get_global_value(store, instance, "count_32")
+      assert {:ok, -64} = Wasmex.Instance.get_global_value(store, instance, "count_64")
```
Comment (tessi) [inline test/wasmex/instance_test.exs:217]: Can you maybe add one more assertion where we test the error when accessing an unsupported type? Like a v128 maybe

**Review by tessi — APPROVED** [review]
Wow, this is great 💜 love the change, fits perfectly to existing conventions, implements a missing festure and reads really well.

Thank you.


If you like, you may add your work to the changelog :)

File: test/wasmex/instance_test.exs (hunk 4 above)
Comment (RoyalIcing) [reply test/wasmex/instance_test.exs:217]: I’ve added asserts for `externref` and `funcref`. It looks like `v128` cannot be used as globals yet.

**Pull request merged**